Framework/Algorithms/src/MultipleScatteringCorrection.cpp

```diff
@@ -0,0 +1,403 @@
+      interpolateLinearInplace(histNew, m_xStep);
+      m_outputWS->setHistogram(i, histNew);
+    }
+
+    prog.report();
+
+    PARALLEL_END_INTERUPT_REGION
+  }
+  PARALLEL_CHECK_INTERUPT_REGION
+
+  g_log.notice() << "finished integration.\n";
+
+  // set the output workspace group
+  // TODO: additional workspace will be added in as we are gradually implementing
+  //       support for container-sample scattering correction
+  const std::string outWSName = getProperty("OutputWorkspace");
+  std::vector<std::string> names;
+  names.emplace_back(outWSName + "_sampleOnly");
+  API::AnalysisDataService::Instance().addOrReplace(names.back(), m_outputWS);
+  // group
+  auto group = createChildAlgorithm("GroupWorkspaces");
+  group->initialize();
+  group->setProperty("InputWorkspaces", names);
+  group->setProperty("OutputWorkspace", outWSName);
+  group->execute();
+  API::WorkspaceGroup_sptr outWS = group->getProperty("OutputWorkspace");
+
+  setProperty("OutputWorkspace", outWS);
+}
+
+/**
+ * @brief parse and assign corresponding values from input properties
+ *
+ */
+void MultipleScatteringCorrection::parseInputs() {
+  // Get input workspace
+  m_inputWS = getProperty("InputWorkspace");
+
+  // Get the beam direction
+  m_beamDirection = m_inputWS->getInstrument()->getBeamDirection();
+
+  // Get the total number of wavelength points, default to use all if not specified
+  m_num_lambda = isDefault("NumberOfWavelengthPoints") ? static_cast<int64_t>(m_inputWS->blocksize())
+                                                       : getProperty("NumberOfWavelengthPoints");
+  // -- while we're here, compute the step in bin number between two adjacent points
+  const auto specSize = static_cast<int64_t>(m_inputWS->blocksize());
+  m_xStep = std::max(int64_t(1), specSize / m_num_lambda); // Bin step between points to calculate
+  // -- notify the user of the bin step
+  std::ostringstream msg;
+  msg << "Numerical integration performed every " << m_xStep << " wavelength points";
+  g_log.information() << msg.str();
+
+  // Get the element size
+  m_elementSize = getProperty("ElementSize"); // in mm
+  m_elementSize = m_elementSize * 1e-3;       // convert to m
+
+  // Get the material
+  const auto &sample = m_inputWS->sample();
+  // -- process the sample
+  m_material = sample.getShape().material();
+  // -- process the sample environment (container)
+  // TODO:
+
+  // Get total scattering cross-section
+  // NOTE: the angstrom^-2 to barns and the angstrom^-1 to cm^-1
+  // will cancel for mu to give units: cm^-1
+  // -- sample
+  m_sampleLinearCoefTotScatt = -m_material.totalScatterXSection() * m_material.numberDensityEffective() * 100;
+  // -- container
+  // TODO:
+
+  // Create output workspace
+  // NOTE: this output workspace is just a Workspace2D of factor that can be applied to the
+  // corresponding EventWorkspace.
+  // Therefore, it is inherently unitless.
+  m_outputWS = create<HistoWorkspace>(*m_inputWS);
```

Comment:
It is slightly misleading that `parseInputs` also sets up the output workspace. Related, it would be better to co-locate this code with the code that sets-up the x-axes (from the exec) as well. I believe that one should [clone](https://github.com/mantidproject/mantid/blob/master/Framework/Algorithms/src/CloneWorkspace.cpp#L46) then `Rebin` based on the other parameters. Then one wouldn't really need the input workspace anymore because all of the sample information is already copied to member variables.

Reply:
It is not clear to me what should be done here (seems like there are several recommendations mangled together here).